node/lib/util/stash_util.js

```diff
@@ -0,0 +1,391 @@
+ * HEAD commits.  If the specified `all` is true, include untracked files in
+ * the stash and clean them.  Do not stash any information for the meta-repo
+ * itself.  Update the `refs/meta-stash` reference and its reflog to point to a
+ * new stash commit.  This commit will have the current HEAD of the repository
+ * as its child, and a tree with containing updated shas for stashed submodules
+ * pointing to their respective stash commits.  In each stashed submodule,
+ * crete a synthetic-meta-ref in the form of `refs/sub-stash/${sha}`, where
+ * `sha` is the stash commit of that submodule.  Return a map from submodule
+ * name to stashed commit for each submodule that was stashed.
+ *
+ * @param {NodeGit.Repository} repo
+ * @param {RepoStatus}         status
+ * @param {Boolean}            all
+ * @return {Object}    submodule name to stashed commit
+ */
+exports.save = co.wrap(function *(repo, status, all) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(status, RepoStatus);
+    assert.isBoolean(all);
+
+    const subResults = {};  // name to sha
+    const subChanges = {};  // name to TreeUtil.Change
+    const subRepos   = {};  // name to submodule open repo
+
+    const sig = repo.defaultSignature();
+
+    // First, we process the submodules.  If a submodule is open and dirty,
+    // we'll create the stash commits in its repo, populate `subResults` with
+    // the `Stash.Submodule` that will be returned, `subChanges` with the sha
+    // of the commit to be made to be used in generating the new submodule
+    // tree, and `subRepos` to cache the open repo for each sub to be used
+    // later.
+
+    const submodules = status.submodules;
+    yield Object.keys(submodules).map(co.wrap(function *(name) {
+        const sub = submodules[name];
+        const wd = sub.workdir;
+        if (null === wd ||
+            (wd.status.isClean() &&
+                (!all || 0 === Object.keys(wd.status.workdir).length))) {
+            // Nothing to do for closed or clean subs
+
+            return;                                                   // RETURN
+        }
+        const subRepo = yield SubmoduleUtil.getRepo(repo, name);
+        subRepos[name] = subRepo;
+        const FLAGS = NodeGit.Stash.FLAGS;
+        const flags = all ? FLAGS.INCLUDE_UNTRACKED : FLAGS.DEFAULT;
+        const stashId = yield NodeGit.Stash.save(subRepo, sig, "stash", flags);
+        subResults[name] = stashId.tostrS();
+
+        // Record the values we've created.
+
+        subChanges[name] = new TreeUtil.Change(
+                                            stashId,
+                                            NodeGit.TreeEntry.FILEMODE.COMMIT);
+    }));
+    const head = yield repo.getHeadCommit();
+    const headTree = yield head.getTree();
+    const subsTree = yield TreeUtil.writeTree(repo, headTree, subChanges);
+    const stashId = yield NodeGit.Commit.create(repo,
```

Comment:
It's normal to me; I find it a little bothersome to write but pleasant to read.

No desire to extend column length at the moment.